### PR TITLE
ast, parser: extract create view parameters to CreateViewStmt

### PR DESCRIPTION
**Pull request by AndrewDi**
As part of view feature implement, extract create view parameters to CreateViewStmt, ref [pingcap#8416](https://github.com/pingcap/tidb/pull/8416)

### CODE REVIEW

**Comment by AndrewDi** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
@AndrewDi Please create an issue in tidb repo to trace the P1 tasks.

File: parser_test.go [hunk 1]
```diff
@@ -2240,6 +2240,22 @@ func (s *testParserSuite) TestView(c *C) {
...
+                  WITH CASCADED CHECK OPTION;`
+
+	var st ast.StmtNode
+	st, err = p.ParseOneStmt(src,"","")
+	c.Assert(err, IsNil)
+	v, ok = st.(*ast.CreateViewStmt)
+	c.Assert(ok, IsTrue)
+	c.Assert(v.OrReplace, IsTrue)
+	c.Assert(v.Algorithm, Equals, ast.AlgorithmUndefined)
+	c.Assert(v.Definer, NotNil)
```
Comment (XuHuaiyu) [inline parser_test.go]: Why not check Equals?

Comment (XuHuaiyu) [reply parser_test.go]: Add check for viewName, colList, view_select_stmt?

File: ast/ddl.go [hunk 2]
```diff
@@ -547,6 +548,52 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
...
+// See https://dev.mysql.com/doc/refman/5.7/en/view-algorithms.html
+type Algorithm int
+
+const (
+	AlgorithmUndefined Algorithm= iota
+	AlgorithmMerge
+	AlgorithmTemptable
+)
+
+var AlgorithmType = map[Algorithm]string {
```
Comment (zz-jason) [inline ast/ddl.go]: how about defining a `String()` function for `Algorithm`:
```go
func (a *Algorithm) String() string {
    switch *a {
        case AlgorithmUndefined:  return "UNDEFINED"
        case AlgorithmMerge    :  return "MERGE"
        case AlgorithmTemptable:  return "TEMPTABLE"
    }
    return "WRONG_ALGORITHM"
}
```

File: ast/ddl.go [hunk 3]
```diff
@@ -547,6 +548,52 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
 	return v.Leave(n)
 }
 
+
+// Algorithm is VIEW's SQL AlGORITHM characteristic
+// See https://dev.mysql.com/doc/refman/5.7/en/view-algorithms.html
+type Algorithm int
```
Comment (zz-jason) [inline ast/ddl.go]: s/Algorithm/ViewAlgorithm/?

File: ast/ddl.go [hunk 4]
```diff
@@ -547,6 +548,52 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
...
+
+var AlgorithmType = map[Algorithm]string {
+	AlgorithmUndefined:	"UNDEFINED",
+	AlgorithmMerge:		"MERGE",
+	AlgorithmTemptable:	"TEMPTABLE",
+}
+
+// Security is VIEW's SQL SECURITY characteristic
+// See https://dev.mysql.com/doc/refman/5.7/en/create-view.html
+type Security int
```
Comment (zz-jason) [inline ast/ddl.go]: s/Security/ViewSecurity/

File: ast/ddl.go [hunk 5]
```diff
@@ -547,6 +548,52 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
...
+
+const (
+	SecurityDefiner Security = iota
+	SecurityInvoker
+)
+
+var SecurityType  = map[Security]string {
```
Comment (zz-jason) [inline ast/ddl.go]: ditto

File: ast/ddl.go [hunk 6]
```diff
@@ -547,6 +548,52 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
...
+	SecurityDefiner:	"DEFINER",
+	SecurityInvoker:	"INVOKER",
+}
+
+
+// CheckOption is VIEW's WITH CHECK OPTION clause part
+// See https://dev.mysql.com/doc/refman/5.7/en/view-check-option.html
+type CheckOption int
```
Comment (zz-jason) [inline ast/ddl.go]: s/CheckOption/ViewCheckOption/

File: ast/ddl.go [hunk 7]
```diff
@@ -547,6 +548,52 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
...
+
+const (
+	CheckOptionLocal CheckOption = iota
+	CheckOptionCascaded
+)
+
+var CheckOptionType = map[CheckOption]string {
```
Comment (zz-jason) [inline ast/ddl.go]: ditto

File: ast/ddl.go [hunk 8]
```diff
@@ -556,12 +603,30 @@ type CreateViewStmt struct {
...
 // Accept implements Node Accept interface.
 func (n *CreateViewStmt) Accept(v Visitor) (Node, bool) {
-	// TODO: implement the details.
-	return n, true
+	newNode, skipChildren := v.Enter(n)
+	if skipChildren {
+		return v.Leave(newNode)
+	}
+	n = newNode.(*CreateViewStmt)
+	node, ok := n.ViewName.Accept(v)
```
Comment (zz-jason) [inline ast/ddl.go]: should we visit `n.ddlNode`?

File: ast/ddl.go (hunk 2 above)
Comment (zz-jason) [reply ast/ddl.go]: Here is the benchmark result:
```
[jianzhang.zj:/tmp]
➜ go test -bench Benchmark -run xx . -count 5
goos: darwin
goarch: amd64
BenchmarkString-8       2000000000               0.62 ns/op
BenchmarkString-8       2000000000               0.63 ns/op
BenchmarkString-8       2000000000               0.63 ns/op
BenchmarkString-8       2000000000               0.62 ns/op
BenchmarkString-8       2000000000               0.62 ns/op
BenchmarkMapAcess-8     500000000                3.75 ns/op
BenchmarkMapAcess-8     500000000                3.79 ns/op
BenchmarkMapAcess-8     500000000                3.75 ns/op
BenchmarkMapAcess-8     500000000                3.74 ns/op
BenchmarkMapAcess-8     500000000                3.75 ns/op
PASS
ok      _/tmp   17.924s
```

Here is the benchmark code:
```go
package main

import "testing"

type Algorithm int

const (
    AlgorithmUndefined Algorithm = iota
    AlgorithmMerge
    AlgorithmTemptable
)

func (a *Algorithm) String() string {
    switch *a {
    case AlgorithmUndefined:
        return "UNDEFINED"
    case AlgorithmMerge:
        return "MERGE"
    case AlgorithmTemptable:
        return "TEMPTABLE"
    }
    return "WRONG_ALGORITHM"
}

var AlgorithmType = map[Algorithm]string{
    AlgorithmUndefined: "UNDEFINED",
    AlgorithmMerge:     "MERGE",
    AlgorithmTemptable: "TEMPTABLE",
}

func BenchmarkString(b *testing.B) {
    var a Algorithm = AlgorithmUndefined
    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        a.String()
    }
}

func BenchmarkMapAcess(b *testing.B) {
    var a Algorithm = AlgorithmUndefined
    b.ResetTimer()
    for i := 0; i < b.N; i++ {
        _ = AlgorithmType[a]
    }
}
```

File: parser_test.go [hunk 9]
```diff
@@ -15,6 +15,7 @@ package parser
 
 import (
 	"fmt"
+	"github.com/pingcap/parser/model"
```
Comment (XuHuaiyu) [inline parser_test.go]: move this line down out of the native packages

File: parser_test.go [hunk 10]
```diff
@@ -2241,19 +2242,24 @@ func (s *testParserSuite) TestView(c *C) {
...
 	c.Assert(ok, IsTrue)
 	c.Assert(v.OrReplace, IsTrue)
 	c.Assert(v.Algorithm, Equals, ast.AlgorithmUndefined)
-	c.Assert(v.Definer, NotNil)
+	c.Assert(v.Definer.Username, Equals, "root")
+	c.Assert(v.Definer.Hostname, Equals, "localhost")
+	c.Assert(v.Cols[0], Equals, model.CIStr{"a", "a"})
+	c.Assert(v.Cols[1], Equals, model.CIStr{"b", "b"})
+	c.Assert(v.Cols[2], Equals, model.CIStr{"c", "c"})
+	c.Assert(strings.TrimSpace(v.Select.Text()), Equals, "select c,d,e from t")
```
Comment (XuHuaiyu) [inline parser_test.go]: Why do we TrimSpace here?
If there are redundant spaces, we need to change parser.y to extract the exact sql text.

File: ast/ddl.go [hunk 11]
```diff
@@ -547,21 +548,99 @@ func (n *TableToTable) Accept(v Visitor) (Node, bool) {
...
+	switch *v {
+	case SecurityDefiner:
+		return "DEFINER"
+	case SecurityInvoker:
+		return "INVOKER"
+	}
+	return "DEFINER"
+}
+
+// ViewCheckOption is VIEW's WITH CHECK OPTION clause part
```
Comment (XuHuaiyu) [inline ast/ddl.go]: Add a `.` at the end of this comment.
So as the other comment lines if they are complete sentences.

File: parser_test.go (hunk 10 above)
Comment (AndrewDi) [reply parser_test.go]: Just done yet...

**Comment by AndrewDi** [comment]
@XuHuaiyu @zz-jason PTAL

File: parser.y [hunk 12]
```diff
@@ -2132,15 +2132,25 @@ CreateViewStmt:
...
+			Definer:       $4.(*auth.UserIdentity),
+			Security:      $5.(ast.ViewSecurity),
 		}
 		if $8 != nil{
 			x.Cols = $8.([]model.CIStr)
 		}
+		if $11 !=nil {
+		    x.CheckOption = $11.(ast.ViewCheckOption)
+		    endOffset := parser.startOffset(&yyS[yypt])
+		    selStmt.SetText(strings.TrimSpace(parser.src[startOffset:endOffset]))
```
Comment (XuHuaiyu) [inline parser.y]: I tested it locally. Seems we do not need this `strings.TrimSpace`.

Comment (AndrewDi) [reply parser.y]: If I remove strings.TrimSpace, extra space would make test fail.
```
FAIL: parser_test.go:2220: testParserSuite.TestView

parser_test.go:2265:
    c.Assert(v.Select.Text(), Equals, "select c,d,e from t")
... obtained string = "" +
...     "select c,d,e from t \n" +
...     "                  "
... expected string = "select c,d,e from t"
```

Comment (AndrewDi) [reply parser.y]: You can use sql string for test:
```go
src := `CREATE OR REPLACE ALGORITHM = UNDEFINED DEFINER = root@localhost
                  SQL SECURITY DEFINER
			      VIEW V(a,b,c) AS select c,d,e from t 
                  WITH CASCADED CHECK OPTION;`
```

**Comment by XuHuaiyu** [comment]
LGTM
PTAL @zz-jason

File: ast/ddl.go (hunk 8 above)
Comment (zz-jason) [reply ast/ddl.go]: got it.

**Pull request merged**